File: packages/flutter/android/src/main/kotlin/technology/breez/spark/BreezSdkSparkPasskeyPlugin.kt
package technology.breez.spark

import android.app.Activity
import android.util.Base64
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import technology.breez.spark.passkey.core.CredentialManagerPrfCore
import technology.breez.spark.passkey.core.CredentialManagerPrfCoreException
import technology.breez.spark.passkey.core.PostCreateGraceTracker

/**
 * Flutter plugin for passkey PRF operations on Android.
 *
 * Thin MethodChannel wrapper around [CredentialManagerPrfCore]. All of the
 * WebAuthn JSON, Credential Manager, and PRF-extraction plumbing lives in
 * the core helper; this file only translates Flutter's method arguments
 * and maps [CredentialManagerPrfCoreException] into MethodChannel error
 * codes understood by the Dart side.
 *
 * Do not hand-edit [CredentialManagerPrfCore.kt] here: it is a generated
 * mirror of the canonical copy under
 * `crates/breez-sdk/bindings/langs/shared/android-passkey/`. Run
 * `cargo xtask sync-passkey-core` after editing the canonical file.
 */
class BreezSdkSparkPasskeyPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null

    /**
     * Plugin-scoped coroutine scope. Cancelled in [onDetachedFromEngine] so
     * any in-flight passkey ceremony does not outlive the plugin and leak
     * the captured Activity. SupervisorJob keeps siblings alive if one
     * branch fails, matching the per-call try/catch pattern below.
     */
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    /**
     * Shared across each per-call [CredentialManagerPrfCore] instance so the
     * post-create grace armed by one ceremony is consumed by the next.
     * Without sharing, every fresh-per-call core would have its own empty
     * tracker and the cross-call grace would never fire.
     */
    private val graceTracker = PostCreateGraceTracker()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "breez_sdk_spark_passkey")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "deriveSeeds" -> handleDeriveSeeds(call, result)
            "createPasskey" -> handleCreatePasskey(call, result)
            "isSupported" -> result.success(CredentialManagerPrfCore.isSupported())
            "checkDomainAssociation" -> handleCheckDomainAssociation(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleDeriveSeeds(call: MethodCall, result: Result) {
        @Suppress("UNCHECKED_CAST")
        val salts = call.argument<List<String>>("salts")
        val rpId = call.argument<String>("rpId")
        val rpName = call.argument<String>("rpName")
        val userName = call.argument<String>("userName")
        val userDisplayName = call.argument<String>("userDisplayName")
        val autoRegister = call.argument<Boolean>("autoRegister") ?: false

        if (salts == null || rpId == null || rpName == null || userName == null || userDisplayName == null) {
            result.error("ERR_PASSKEY", "Invalid arguments", null)
            return
        }
        val currentActivity = activity ?: run {
            result.error("ERR_PASSKEY", "No activity available", null)
            return
        }

        // Caller-supplied allow-list. The host passes any known
        // credential IDs from the Dart side before the MethodChannel
        // call. The native plugin never reads or writes credential IDs
        // itself.
        val allowIds: List<ByteArray> =
            (call.argument<List<String>>("allowCredentials") ?: emptyList()).map {
                Base64.decode(it, Base64.NO_WRAP)
            }
        val preferImmediate = call.argument<Boolean>("preferImmediatelyAvailableCredentials")

        scope.launch {
            try {
                val derivation = CredentialManagerPrfCore(
                    rpId = rpId,
                    rpName = rpName,
                    userName = userName,
                    userDisplayName = userDisplayName,
                    activityProvider = { currentActivity },
                    graceTracker = graceTracker,
                ).deriveSeeds(
                    salts = salts,
                    autoRegister = autoRegister,
                    allowCredentials = allowIds,
                    preferImmediatelyAvailableCredentials = preferImmediate ?: true,
                )
                // Encode each seed as base64 so MethodChannel can carry it.
                // The asserted credential ID rides alongside (null when no
                // assertion ran) so the Dart side surfaces it via
                // DeriveSeedsOutput, matching the other bindings.
                result.success(mapOf(
                    "seeds" to derivation.seeds.map { Base64.encodeToString(it, Base64.NO_WRAP) },
                    "credentialId" to derivation.credentialId?.let { Base64.encodeToString(it, Base64.NO_WRAP) },
                ))
            } catch (e: CredentialManagerPrfCoreException) {
                result.error(e.errorCode, e.message ?: e.defaultMessage, null)
            } catch (e: Exception) {
                result.error("ERR_PASSKEY", e.message ?: e.toString(), null)
            }
        }
    }

    private fun handleCheckDomainAssociation(call: MethodCall, result: Result) {
        val rpId = call.argument<String>("rpId")
        if (rpId == null) {
            result.error("ERR_PASSKEY", "Invalid arguments", null)
            return
        }
        val currentActivity = activity ?: run {
            result.error("ERR_PASSKEY", "No activity available", null)
            return
        }
        scope.launch {
            try {
                // Branding fields are unused by the domain check; pass
                // rpId as a placeholder since this is a check-only core.
                val outcome = CredentialManagerPrfCore(
                    rpId = rpId,
                    rpName = rpId,
                    userName = rpId,
                    userDisplayName = rpId,
                    activityProvider = { currentActivity },
                ).checkDomainAssociation()
                // Soft-fail to Skipped on Android: see the upstream
                // PasskeyProvider for the rationale (CredentialManager
                // runs its own DAL check internally and a public-API
                // mismatch can be a stale-cache false negative).
                result.success(when (outcome) {
                    is technology.breez.spark.passkey.core.DomainAssociationResult.Associated ->
                        mapOf("kind" to "Associated")
                    is technology.breez.spark.passkey.core.DomainAssociationResult.NotAssociated ->
                        mapOf(
                            "kind" to "Skipped",
                            "reason" to "[soft-fail on Android] ${outcome.reason}",
                        )
                    is technology.breez.spark.passkey.core.DomainAssociationResult.Skipped ->
                        mapOf("kind" to "Skipped", "reason" to outcome.reason)
                })
            } catch (e: Exception) {
                result.success(mapOf(
                    "kind" to "Skipped",
                    "reason" to "Domain association probe failed: ${e.message ?: e.toString()}",
                ))
            }
        }
    }

    private fun handleCreatePasskey(call: MethodCall, result: Result) {
        val rpId = call.argument<String>("rpId")
        val rpName = call.argument<String>("rpName")
        val userName = call.argument<String>("userName")
        val userDisplayName = call.argument<String>("userDisplayName")

        if (rpId == null || rpName == null || userName == null || userDisplayName == null) {
            result.error("ERR_PASSKEY", "Invalid arguments", null)
            return
        }
        val currentActivity = activity ?: run {
            result.error("ERR_PASSKEY", "No activity available", null)
            return
        }

        val excludeIds: List<ByteArray> =
            (call.argument<List<String>>("excludeCredentials") ?: emptyList()).map {
                Base64.decode(it, Base64.NO_WRAP)
            }

        scope.launch {
            try {
                val credential = CredentialManagerPrfCore(
                    rpId = rpId,
                    rpName = rpName,
                    userName = userName,
                    userDisplayName = userDisplayName,
                    activityProvider = { currentActivity },
                    graceTracker = graceTracker,
                ).register(excludeIds)
                // The core's `register` arms the shared grace tracker so the
                // next `deriveSeeds` call holds out the credential's
                // PRF-readiness window without the wrapper having to.
                result.success(mapOf(
                    "credentialId" to Base64.encodeToString(credential.credentialId, Base64.NO_WRAP),
                    "userId" to Base64.encodeToString(credential.userId, Base64.NO_WRAP),
                    "aaguid" to credential.aaguid?.let { Base64.encodeToString(it, Base64.NO_WRAP) },
                    "backupEligible" to credential.backupEligible,
                ))
            } catch (e: CredentialManagerPrfCoreException) {
                result.error(e.errorCode, e.message ?: e.defaultMessage, null)
            } catch (e: Exception) {
                result.error("ERR_PASSKEY", e.message ?: e.toString(), null)
            }
        }
    }

    private val CredentialManagerPrfCoreException.errorCode: String
        get() = when (kind) {
            CredentialManagerPrfCore.Kind.UserCancelled -> "ERR_USER_CANCELLED"
            CredentialManagerPrfCore.Kind.UserTimedOut -> "ERR_USER_TIMED_OUT"
            CredentialManagerPrfCore.Kind.CredentialNotFound -> "ERR_NO_CREDENTIAL"
            CredentialManagerPrfCore.Kind.PrfNotSupported -> "ERR_PRF_NOT_SUPPORTED"
            CredentialManagerPrfCore.Kind.Configuration -> "ERR_CONFIGURATION"
            CredentialManagerPrfCore.Kind.CredentialAlreadyExists -> "ERR_CREDENTIAL_ALREADY_EXISTS"
            else -> "ERR_PASSKEY"
        }

    private val CredentialManagerPrfCoreException.defaultMessage: String
        get() = when (kind) {
            CredentialManagerPrfCore.Kind.UserCancelled -> "User cancelled the passkey operation"
            CredentialManagerPrfCore.Kind.UserTimedOut -> "Authenticator timed out"
            CredentialManagerPrfCore.Kind.CredentialNotFound -> "No passkey credential found for this domain"
            CredentialManagerPrfCore.Kind.PrfNotSupported -> "PRF not supported by authenticator"
            CredentialManagerPrfCore.Kind.AuthenticationFailed -> "Passkey authentication failed"
            CredentialManagerPrfCore.Kind.PrfEvaluationFailed -> "PRF evaluation failed"
            CredentialManagerPrfCore.Kind.Configuration -> "Platform or app configuration error"
            CredentialManagerPrfCore.Kind.CredentialAlreadyExists -> "A passkey for this app already exists on this device"
            CredentialManagerPrfCore.Kind.Generic -> "Passkey operation failed"
        }
}

File: packages/flutter/ios/Classes/BreezSdkSparkPasskeyPlugin.swift
import AuthenticationServices
import Flutter
import UIKit

/// Flutter plugin for passkey PRF operations on iOS.
///
/// Thin MethodChannel bridge over the shared `PasskeyAssertionCore`.
/// Behavioral parity with the upstream Swift `PasskeyProvider` and the
/// React Native module is enforced by routing all ASAuthorizationController
/// orchestration through the same canonical core.
@available(iOS 18.0, *)
public class BreezSdkSparkPasskeyPlugin: NSObject, FlutterPlugin {

    /// Shared across the per-call cores so the post-create PRF-readiness
    /// grace armed by `createPasskey` survives to an immediately-following
    /// `deriveSeeds` (sparing a second prompt). The registrar retains this
    /// plugin instance, so it persists across calls; mirrors the Android plugin.
    private let graceTracker = PostCreateGraceTracker()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "breez_sdk_spark_passkey",
            binaryMessenger: registrar.messenger()
        )
        let instance = BreezSdkSparkPasskeyPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "createPasskey":
            handleCreatePasskey(call: call, result: result)
        case "deriveSeeds":
            handleDeriveSeeds(call: call, result: result)
        case "isSupported":
            if #available(iOS 18.0, *) {
                result(true)
            } else {
                result(false)
            }
        case "checkDomainAssociation":
            handleCheckDomainAssociation(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Method handlers

    private func handleCreatePasskey(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let rpId = args["rpId"] as? String,
              let rpName = args["rpName"] as? String,
              let userName = args["userName"] as? String,
              let userDisplayName = args["userDisplayName"] as? String
        else {
            result(FlutterError(code: "ERR_PASSKEY", message: "Invalid arguments", details: nil))
            return
        }

        var excludeCredentials: [Data] = []
        if let rawIds = args["excludeCredentials"] as? [FlutterStandardTypedData] {
            excludeCredentials = rawIds.map { $0.data }
        } else if let base64Ids = args["excludeCredentials"] as? [String] {
            excludeCredentials = base64Ids.compactMap { Data(base64Encoded: $0) }
        }

        let core = PasskeyAssertionCore(
            rpId: rpId, rpName: rpName, userName: userName, userDisplayName: userDisplayName,
            graceTracker: graceTracker
        )
        Task { @MainActor in
            do {
                let registered = try await core.register(excludeCredentials: excludeCredentials)
                result([
                    "credentialId": registered.credentialId.base64EncodedString(),
                    "userId": registered.userId.base64EncodedString(),
                    "aaguid": registered.aaguid?.base64EncodedString() as Any?,
                    "backupEligible": registered.backupEligible as Any?,
                ])
            } catch let err as PasskeyAssertionError {
                result(Self.flutterError(from: err))
            } catch {
                result(FlutterError(code: "ERR_PASSKEY", message: error.localizedDescription, details: nil))
            }
        }
    }

    private func handleDeriveSeeds(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let salts = args["salts"] as? [String],
              let rpId = args["rpId"] as? String,
              let rpName = args["rpName"] as? String,
              let userName = args["userName"] as? String,
              let userDisplayName = args["userDisplayName"] as? String
        else {
            result(FlutterError(code: "ERR_PASSKEY", message: "Invalid arguments", details: nil))
            return
        }

        let autoRegister = (args["autoRegister"] as? Bool) ?? false

        // Encode salts as UTF-8 bytes; the OS uses these directly as
        // PRF eval inputs (saltInput1, saltInput2).
        let saltDatas: [Data] = salts.compactMap { $0.data(using: .utf8) }
        guard saltDatas.count == salts.count else {
            result(FlutterError(code: "ERR_PASSKEY", message: "Failed to encode salts as UTF-8", details: nil))
            return
        }

        var allowCredentials: [Data] = []
        if let base64Ids = args["allowCredentials"] as? [String] {
            allowCredentials = base64Ids.compactMap { Data(base64Encoded: $0) }
        } else if let rawIds = args["allowCredentials"] as? [FlutterStandardTypedData] {
            allowCredentials = rawIds.map { $0.data }
        }
        let preferImmediate = args["preferImmediatelyAvailableCredentials"] as? Bool

        let core = PasskeyAssertionCore(
            rpId: rpId, rpName: rpName, userName: userName, userDisplayName: userDisplayName,
            graceTracker: graceTracker
        )
        Task { @MainActor in
            do {
                let derivation = try await core.deriveSeeds(
                    salts: saltDatas,
                    autoRegister: autoRegister,
                    allowCredentials: allowCredentials,
                    preferImmediatelyAvailableCredentials: preferImmediate ?? true
                )
                result([
                    "seeds": derivation.seeds.map { $0.base64EncodedString() },
                    "credentialId": derivation.credentialId.map { $0.base64EncodedString() } ?? NSNull(),
                ])
            } catch let err as PasskeyAssertionError {
                result(Self.flutterError(from: err))
            } catch {
                result(FlutterError(code: "ERR_PASSKEY", message: error.localizedDescription, details: nil))
            }
        }
    }

    private func handleCheckDomainAssociation(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let rpId = args["rpId"] as? String
        else {
            result(FlutterError(code: "ERR_PASSKEY", message: "Invalid arguments", details: nil))
            return
        }
        // Optional explicit team ID: most callers leave it nil and let
        // the canonical core auto-detect from the running app's
        // signing info. Useful for unit tests / sandboxed contexts
        // where SecTask / provisioning-profile lookup doesn't work.
        let explicitTeamId = args["teamId"] as? String

        // Branding fields are unused by the domain check; pass rpId as
        // a placeholder since this is a check-only core.
        let core = PasskeyAssertionCore(
            rpId: rpId, rpName: rpId, userName: rpId, userDisplayName: rpId,
            explicitTeamId: explicitTeamId
        )
        Task { @MainActor in
            let outcome = await core.checkDomainAssociation()
            result(Self.serializeDomainAssociation(outcome))
        }
    }

    /// Serialize `IosDomainAssociation` into the JSON-shaped map the
    /// Dart side expects. Mirrors the typed `DomainAssociation`
    /// sealed class one-to-one.
    private static func serializeDomainAssociation(
        _ outcome: IosDomainAssociation
    ) -> [String: Any] {
        switch outcome {
        case .associated:
            return ["kind": "Associated"]
        case .notAssociated(let source, let reason):
            return ["kind": "NotAssociated", "source": source, "reason": reason]
        case .skipped(let reason):
            return ["kind": "Skipped", "reason": reason]
        }
    }

    // MARK: - Error mapping

    private static func flutterError(from err: PasskeyAssertionError) -> FlutterError {
        switch err {
        case .userCancelled:
            return FlutterError(code: "ERR_USER_CANCELLED", message: "User cancelled authentication", details: nil)
        case .userTimedOut:
            return FlutterError(code: "ERR_USER_TIMED_OUT", message: "Authenticator timed out", details: nil)
        case .credentialNotFound(let msg):
            return FlutterError(code: "ERR_NO_CREDENTIAL", message: msg, details: nil)
        case .credentialAlreadyExists(let msg):
            return FlutterError(code: "ERR_CREDENTIAL_ALREADY_EXISTS", message: msg, details: nil)
        case .prfNotSupported:
            return FlutterError(code: "ERR_PRF_NOT_SUPPORTED", message: "PRF not supported by authenticator", details: nil)
        case .prfEvaluationFailed(let msg):
            return FlutterError(code: "ERR_PRF_NOT_SUPPORTED", message: msg, details: nil)
        case .configuration(let msg):
            return FlutterError(code: "ERR_CONFIGURATION", message: msg, details: nil)
        case .authenticationFailed(let msg):
            return FlutterError(code: "ERR_PASSKEY", message: msg, details: nil)
        case .generic(let msg):
            return FlutterError(code: "ERR_PASSKEY", message: msg, details: nil)
        }
    }
}
